File: jdroid-android-firebase-remoteconfig/src/debug/java/com/jdroid/android/firebase/remoteconfig/FirebaseRemoteConfigFragment.java
package com.jdroid.android.firebase.remoteconfig;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.TextView;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.jdroid.android.application.AbstractApplication;
import com.jdroid.android.application.AppModule;
import com.jdroid.android.recycler.AbstractRecyclerFragment;
import com.jdroid.android.recycler.RecyclerViewAdapter;
import com.jdroid.android.recycler.RecyclerViewContainer;
import com.jdroid.android.recycler.RecyclerViewType;
import com.jdroid.java.collections.Lists;
import com.jdroid.java.date.DateUtils;
import com.jdroid.java.remoteconfig.RemoteConfigParameter;
import com.jdroid.java.utils.StringUtils;

import java.util.Date;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class FirebaseRemoteConfigFragment extends AbstractRecyclerFragment {

	@Override
	public void onViewCreated(View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		List<Object> items = Lists.newArrayList();
		items.add("");
		for (AppModule appModule : AbstractApplication.get().getAppModules()) {
			items.addAll(appModule.getRemoteConfigParameters());
		}
		items.addAll(AbstractApplication.get().getDebugContext().getRemoteConfigParameters());
		List<RecyclerViewType> recyclerViewTypes = Lists.newArrayList(new HeaderRecyclerViewType(), new RemoteConfigParameterRecyclerViewType());
		setAdapter(new RecyclerViewAdapter(recyclerViewTypes, items));
	}

	public class HeaderRecyclerViewType extends RecyclerViewType<String, HeaderViewHolder> {

		@Override
		protected Integer getLayoutResourceId() {
			return R.layout.jdroid_firebase_remote_config_header;
		}

		@Override
		protected Class<String> getItemClass() {
			return String.class;
		}

		@Override
		public RecyclerView.ViewHolder createViewHolderFromView(View view) {
			HeaderViewHolder viewHolder = new HeaderViewHolder(view);
			viewHolder.fetchTimeMillis = findView(view, R.id.fetchTimeMillis);
			viewHolder.lastFetchStatus = findView(view, R.id.lastFetchStatus);
			viewHolder.fetch = findView(view, R.id.fetch);
			viewHolder.save = findView(view, R.id.save);
			viewHolder.mocksEnabled = findView(view, R.id.mocksEnabled);
			return viewHolder;
		}

		@Override
		public void fillHolderFromItem(String item, HeaderViewHolder holder) {
			holder.fetch.setEnabled(!MockRemoteConfigLoader.isMocksEnabled());
			holder.fetch.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View view) {
					FirebaseRemoteConfigLoader.get().fetch(false, new OnSuccessListener<Void>() {
						@Override
						public void onSuccess(Void aVoid) {
							executeOnUIThread(new Runnable() {
								@Override
								public void run() {
									getRecyclerViewAdapter().notifyDataSetChanged();
								}
							});
						}
					});
				}
			});

			holder.mocksEnabled.setOnCheckedChangeListener(null);
			holder.mocksEnabled.setChecked(MockRemoteConfigLoader.isMocksEnabled());
			holder.mocksEnabled.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
				@Override
				public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
					MockRemoteConfigLoader.setMocksEnabled(isChecked);
					getRecyclerViewAdapter().notifyDataSetChanged();
				}
			});

			if (!MockRemoteConfigLoader.isMocksEnabled()) {
				FirebaseRemoteConfig firebaseRemoteConfig = FirebaseRemoteConfigLoader.get().getFirebaseRemoteConfig();
				if (firebaseRemoteConfig != null) {
					String fetchDate = DateUtils.INSTANCE.formatDateTime(new Date(firebaseRemoteConfig.getInfo().getFetchTimeMillis()));
					holder.fetchTimeMillis.setText(getString(R.string.jdroid_firebaseRemoteConfigFetchDate, fetchDate));
					int status = firebaseRemoteConfig.getInfo().getLastFetchStatus();
					String statusValue = "Unknown";
					if (status == -1) {
						statusValue = "Success";
					} else if (status == 0) {
						statusValue = "No fetch yet";
					} else if (status == 1) {
						statusValue = "Failure";
					} else if (status == 2) {
						statusValue = "Throttled";
					}
					holder.lastFetchStatus.setText(getString(R.string.jdroid_firebaseRemoteConfigLastFetchStatus, statusValue));
				}
			}
		}

		@NonNull
		@Override
		public RecyclerViewContainer getRecyclerViewContainer() {
			return FirebaseRemoteConfigFragment.this;
		}
	}

	public static class HeaderViewHolder extends RecyclerView.ViewHolder {

		public TextView fetchTimeMillis;
		public TextView lastFetchStatus;
		public Button fetch;
		public Button save;
		public CheckBox mocksEnabled;

		private HeaderViewHolder(View itemView) {
			super(itemView);
		}
	}

	public class RemoteConfigParameterRecyclerViewType extends RecyclerViewType<RemoteConfigParameter, RemoteConfigParameterItemHolder> {

		@Override
		protected Class<RemoteConfigParameter> getItemClass() {
			return RemoteConfigParameter.class;
		}

		@Override
		protected Integer getLayoutResourceId() {
			return R.layout.jdroid_firebase_remote_config_param;
		}

		@Override
		public RecyclerView.ViewHolder createViewHolderFromView(View view) {
			RemoteConfigParameterItemHolder holder = new RemoteConfigParameterItemHolder(view);
			holder.key = findView(view, R.id.key);
			holder.specs = findView(view, R.id.specs);
			holder.source = findView(view, R.id.source);
			holder.value = findView(view, R.id.value);
			holder.save = findView(view, R.id.save);
			return holder;
		}

		@Override
		public void fillHolderFromItem(final RemoteConfigParameter item, final RemoteConfigParameterItemHolder holder) {
			holder.key.setText(item.getKey());
			holder.specs.setText(getString(R.string.jdroid_firebaseRemoteConfigSpec, item.getDefaultValue()));

			if (MockRemoteConfigLoader.isMocksEnabled()) {
				holder.source.setVisibility(View.GONE);
			} else {
				holder.source.setText(getString(R.string.jdroid_firebaseRemoteConfigSource, FirebaseRemoteConfigLoader.get().getSourceName(item)));
				holder.source.setVisibility(View.VISIBLE);
			}

			holder.value.setText(AbstractApplication.get().getRemoteConfigLoader().getString(item));
			holder.value.setEnabled(MockRemoteConfigLoader.isMocksEnabled());

			holder.save.setVisibility(MockRemoteConfigLoader.isMocksEnabled() ? View.VISIBLE : View.GONE);
			holder.save.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					MockRemoteConfigLoader.get().saveRemoteConfigParameter(item, StringUtils.getNotEmptyString(holder.value.getText().toString()));
				}
			});
		}

		@Override
		public Boolean matchViewType(Object item) {
			return item instanceof RemoteConfigParameter;
		}

		@NonNull
		@Override
		public AbstractRecyclerFragment getRecyclerViewContainer() {
			return FirebaseRemoteConfigFragment.this;
		}
	}

	public static class RemoteConfigParameterItemHolder extends RecyclerView.ViewHolder {

		protected TextView key;
		protected TextView specs;
		protected TextView source;
		protected EditText value;
		protected Button save;

		private RemoteConfigParameterItemHolder(View itemView) {
			super(itemView);
		}
	}
}

File: jdroid-android-firebase-remoteconfig/src/debug/java/com/jdroid/android/firebase/remoteconfig/FirebaseRemoteConfigFragment.kt
package com.jdroid.android.firebase.remoteconfig

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.jdroid.android.application.AbstractApplication
import com.jdroid.android.recycler.AbstractRecyclerFragment
import com.jdroid.android.recycler.RecyclerViewAdapter
import com.jdroid.android.recycler.RecyclerViewContainer
import com.jdroid.android.recycler.RecyclerViewType
import com.jdroid.java.date.DateUtils
import com.jdroid.java.remoteconfig.RemoteConfigParameter
import com.jdroid.java.utils.StringUtils
import java.util.Date

class FirebaseRemoteConfigFragment : AbstractRecyclerFragment() {

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val items = mutableListOf<Any>()
        items.add("")
        for (appModule in AbstractApplication.get().appModules) {
            items.addAll(appModule.getRemoteConfigParameters())
        }
        items.addAll(AbstractApplication.get().debugContext.getRemoteConfigParameters())
        val recyclerViewTypes = listOf(HeaderRecyclerViewType(), RemoteConfigParameterRecyclerViewType())
        setAdapter(RecyclerViewAdapter(recyclerViewTypes, items))
    }

    inner class HeaderRecyclerViewType : RecyclerViewType<String, HeaderViewHolder>() {

        override fun getLayoutResourceId(): Int {
            return R.layout.jdroid_firebase_remote_config_header
        }

        override fun getItemClass(): Class<String> {
            return String::class.java
        }

        override fun createViewHolderFromView(view: View): RecyclerView.ViewHolder {
            val viewHolder = HeaderViewHolder(view)
            viewHolder.fetchTimeMillis = findView(view, R.id.fetchTimeMillis)
            viewHolder.lastFetchStatus = findView(view, R.id.lastFetchStatus)
            viewHolder.fetch = findView(view, R.id.fetch)
            viewHolder.mocksEnabled = findView(view, R.id.mocksEnabled)
            return viewHolder
        }

        override fun fillHolderFromItem(item: String, holder: HeaderViewHolder) {
            holder.fetch.isEnabled = !MockRemoteConfigLoader.isMocksEnabled()
            holder.fetch.setOnClickListener {
                FirebaseRemoteConfigLoader.get()
                    .fetch(false) { executeOnUIThread(Runnable { getRecyclerViewAdapter().notifyDataSetChanged() }) }
            }

            holder.mocksEnabled.setOnCheckedChangeListener(null)
            holder.mocksEnabled.isChecked = MockRemoteConfigLoader.isMocksEnabled()
            holder.mocksEnabled.setOnCheckedChangeListener { _, isChecked ->
                MockRemoteConfigLoader.setMocksEnabled(isChecked)
                getRecyclerViewAdapter().notifyDataSetChanged()
            }

            if (!MockRemoteConfigLoader.isMocksEnabled()) {
                val firebaseRemoteConfig = FirebaseRemoteConfigLoader.get().firebaseRemoteConfig
                if (firebaseRemoteConfig != null) {
                    val fetchDate = DateUtils.formatDateTime(Date(firebaseRemoteConfig.info.fetchTimeMillis))
                    holder.fetchTimeMillis.text = getString(R.string.jdroid_firebaseRemoteConfigFetchDate, fetchDate)
                    val status = firebaseRemoteConfig.info.lastFetchStatus
                    var statusValue = "Unknown"
                    if (status == -1) {
                        statusValue = "Success"
                    } else if (status == 0) {
                        statusValue = "No fetch yet"
                    } else if (status == 1) {
                        statusValue = "Failure"
                    } else if (status == 2) {
                        statusValue = "Throttled"
                    }
                    holder.lastFetchStatus.text = getString(R.string.jdroid_firebaseRemoteConfigLastFetchStatus, statusValue)
                }
            }
        }

        override fun getRecyclerViewContainer(): RecyclerViewContainer {
            return this@FirebaseRemoteConfigFragment
        }
    }

    class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        lateinit var fetchTimeMillis: TextView
        lateinit var lastFetchStatus: TextView
        lateinit var fetch: Button
        lateinit var mocksEnabled: CheckBox
    }

    inner class RemoteConfigParameterRecyclerViewType :
        RecyclerViewType<RemoteConfigParameter, RemoteConfigParameterItemHolder>() {

        override fun getItemClass(): Class<RemoteConfigParameter> {
            return RemoteConfigParameter::class.java
        }

        override fun getLayoutResourceId(): Int {
            return R.layout.jdroid_firebase_remote_config_param
        }

        override fun createViewHolderFromView(view: View): RecyclerView.ViewHolder {
            val holder = RemoteConfigParameterItemHolder(view)
            holder.key = findView(view, R.id.key)
            holder.specs = findView(view, R.id.specs)
            holder.source = findView(view, R.id.source)
            holder.value = findView(view, R.id.value)
            holder.save = findView(view, R.id.save)
            return holder
        }

        override fun fillHolderFromItem(item: RemoteConfigParameter, holder: RemoteConfigParameterItemHolder) {
            holder.key.text = item.getKey()
            holder.specs.text = getString(R.string.jdroid_firebaseRemoteConfigSpec, item.getDefaultValue())

            if (MockRemoteConfigLoader.isMocksEnabled()) {
                holder.source.visibility = View.GONE
            } else {
                holder.source.text = getString(R.string.jdroid_firebaseRemoteConfigSource, FirebaseRemoteConfigLoader.get().getSourceName(item))
                holder.source.visibility = View.VISIBLE
            }

            holder.value.setText(AbstractApplication.get().remoteConfigLoader!!.getString(item))
            holder.value.isEnabled = MockRemoteConfigLoader.isMocksEnabled()

            holder.save.visibility = if (MockRemoteConfigLoader.isMocksEnabled()) View.VISIBLE else View.GONE
            holder.save.setOnClickListener {
                MockRemoteConfigLoader.get().saveRemoteConfigParameter(item, StringUtils.getNotEmptyString(holder.value.text.toString()))
            }
        }

        override fun matchViewType(item: Any): Boolean? {
            return item is RemoteConfigParameter
        }

        override fun getRecyclerViewContainer(): AbstractRecyclerFragment {
            return this@FirebaseRemoteConfigFragment
        }
    }

    class RemoteConfigParameterItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        lateinit var key: TextView
        lateinit var specs: TextView
        lateinit var source: TextView
        lateinit var value: EditText
        lateinit var save: Button
    }
}
